### PR TITLE
Increase or Decrease polling rate for instance in a final state ready or failed.

### PR DESCRIPTION
**Pull request by santiago-ventura**
In order to prevent the Operator hits the CF rates limits, we have introduced two new annotations, `service-operator.cf.cs.sap.com/polling-interval-ready` and  `service-operator.cf.cs.sap.com/polling-interval-fail` to facilitate the increase or decrease of the re-queue time duration when a Custom Resources is in a State final Ready or Failed.

### CODE REVIEW

File: website/content/en/docs/tutorials/annotations.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+```yaml
+apiVersion: cf.cs.sap.com/v1alpha1
+kind: ServiceInstance
+  metadata:
+    annotations:
+      service-operator.cf.cs.sap.com/polling-interval-fail: "5m"
+```
+
+In the example above the custom resource will be reconcile every three hours after reaching the final state Failed.
```
Comment (RalfHammer) [inline website/content/en/docs/tutorials/annotations.md:48]:
```suggestion
In the example above the custom resource will be reconcile every five minutes after reaching the final state Failed.
```

File: website/content/en/docs/tutorials/annotations.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  metadata:
+    annotations:
+      service-operator.cf.cs.sap.com/polling-interval-ready: "3h"
+```
+
+In the example above the custom resource will be reconcile every three hours after reaching the state Ready.
+
+## Annotation Polling Interval Fail
+
+The AnnotationPollingIntervalFail annotation is used to specify the duration of the requeue interval at which the operator polls the status of a Custom Resource after the final states Creation Failed and Deletion Failed. Currently it is possible to apply this annotations to ServiceInstance custom resource only.
```
Comment (RalfHammer) [inline website/content/en/docs/tutorials/annotations.md:32]: Why do you not support the other resources as well?

File: internal/controllers/servicebinding_controller.go [hunk 3]
```diff
@@ -317,7 +317,7 @@ func (r *ServiceBindingReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 				return ctrl.Result{RequeueAfter: 10 * time.Minute}, nil
 			}
 			// TODO: apply some increasing period, depending on the age of the last update
-			return ctrl.Result{RequeueAfter: 10 * time.Minute}, nil
+			return getPollingInterval(serviceInstance.GetAnnotations(), "10m", cfv1alpha1.AnnotationPollingIntervalReady), nil
```
Comment (RalfHammer) [inline internal/controllers/servicebinding_controller.go:320]: You should read here from serviceBinding instead?

Comment (santiago-ventura) [reply internal/controllers/servicebinding_controller.go]: Right :-)

File: website/content/en/docs/tutorials/annotations.md (hunk 2 above)
Comment (santiago-ventura) [reply website/content/en/docs/tutorials/annotations.md:32]: @RalfHammer:  
We have technical debt with the Service Binding Controller. We haven't implemented the HandleError function or the MaxRetries capabilities in this controller, so the AnnotationPollingIntervalFail has not been applied yet.

File: internal/controllers/reconciler_helpers.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+import (
+	"strconv"
+	"time"
+
+	"github.com/go-logr/logr"
+	cfv1alpha1 "github.com/sap/cf-service-operator/api/v1alpha1"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+// IncrementRetryCounterAndCheckRetryLimit increments the retry counter for a ServiceInstance and checks if the number of retries has exceeded the maximum allowed retries.
```
Comment (RalfHammer) [inline internal/controllers/reconciler_helpers.go:17]: you should remove these old comments

File: internal/controllers/reconciler_helpers_test.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	cfv1alpha1 "github.com/sap/cf-service-operator/api/v1alpha1"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+func TestReconcilerHelpers(t *testing.T) {
+	RegisterFailHandler(Fail)
+	// RunSpecs(t, "Reconciler Helpers Suite")
```
Comment (RalfHammer) [inline internal/controllers/reconciler_helpers_test.go:22]: Do you want to activate this test?

**Review by RalfHammer — DISMISSED** [review]

**Comment by santiago-ventura** [comment]
Hi @TheBigElmo, please review the PR.
Ralf has already completed his review.
Thxs!

**Comment by TheBigElmo** [comment]
Why would you set this on a CRO level, but not in general on startup of the controller, so you can override the default values? Do you have numbers for end users, when to adjust it (e.g. number of resources)? Are the CF rate limits maybe too low?

**Comment by santiago-ventura** [comment]
Hi @TheBigElmo 

I understand your question about why not set these time durations for general use on the controller's startup.

In our use case, we might want different timeout values depending on which instance we're creating and its purpose.

So, by making it configurable per instance, we can always set it when we create the resource.

We can differentiate it, **so we took this approach because of the ability to decouple it per resource.**

**Comment by TheBigElmo** [comment]
That's completely fine to do this on a CRO level, but it should have a possibility on startup too.

**Comment by RalfHammer** [comment]
>That's completely fine to do this on a CRO level, but it should have a possibility on startup too.
We would prefer to implement this possibility on startup in a separate PR.

**Comment by santiago-ventura** [comment]
@TheBigElmo, you are right; it is possible to have both options.
We have more backlog items and can introduce that feature in another PR next.

**Comment by TheBigElmo** [comment]
ok, let's do it with another PR next.
What about the questions: 

> Do you have numbers for end users, when to adjust it (e.g. number of resources)? Are the CF rate limits maybe too low?

File: website/content/en/docs/tutorials/annotations.md
[the remark this answers is not in the document]
Comment (santiago-ventura) [reply website/content/en/docs/tutorials/annotations.md:1]: I made some changes to the document.
This includes information about the default values and behavior under **Default Requeue After Interval**.

**Comment by santiago-ventura** [comment]
> ok, let's do it with another PR next. What about the questions:
> 
> > Do you have numbers for end users, when to adjust it (e.g. number of resources)? Are the CF rate limits maybe too low?

Here is the number we have managed so far without reconciling the limit. It reconciles every 10 min at least; we would have with caching six requests per instance per hour. Keep in mind that Spaces are much more wasteful, taking 240 requests per hour.
 
So, in theory, we would have problems at around 10k instances per user, assuming no other workload. If you add keys, it becomes more like 5k.
 
In practice, we have problems around 200 spaces, which makes sense given that it's 200*240 > 40k requests.
 
We currently have a backlog item for testing how many requests the operator will execute with 10.000 CRs.

**Review by TheBigElmo — APPROVED** [review]
lgtm, please create an additional pr for the controller flags.

**Comment by santiago-ventura** [comment]
Hi @TheBigElmo 
Thank you!

Yes, our product owner will create a backlog item for the controller flags.
We currently have other changes in the making. We will work on controller flags in the following sprints 
to have the CF service operator in production as soon as possible.

**Pull request merged**